src/ports/postgres/modules/knn/knn.sql_in

```diff
@@ -0,0 +1,126 @@
+m4_ifdef(`__HAS_FUNCTION_PROPERTIES__', `READS SQL DATA', `');
+
+
+
+CREATE OR REPLACE FUNCTION MADLIB_SCHEMA.knn(
+    point_source VARCHAR,
+    point_column_name VARCHAR,
+    label_column_name VARCHAR,
+    test_source VARCHAR,
+    test_column_name VARCHAR,
+    id_column_name VARCHAR,
+    operation VARCHAR,
+    k INTEGER
+) RETURNS VARCHAR AS $$
+DECLARE
+    class_test_source REGCLASS;
+    class_point_source REGCLASS;
+    l FLOAT;
+    id INTEGER;
+    vector DOUBLE PRECISION[];
+    cur_pid integer;
+    theResult MADLIB_SCHEMA.knn_result;
+    r MADLIB_SCHEMA.test_table_spec;
+    oldClientMinMessages VARCHAR;
+    returnstring VARCHAR;
+BEGIN
+    oldClientMinMessages :=
+        (SELECT setting FROM pg_settings WHERE name = 'client_min_messages');
+    EXECUTE 'SET client_min_messages TO warning';
+    PERFORM MADLIB_SCHEMA.__knn_validate_src(test_source);
+    PERFORM MADLIB_SCHEMA.__knn_validate_src(point_source);
+    class_test_source := test_source;
+    class_point_source := point_source;
+    --checks
+    IF (k <= 0) THEN
+        RAISE EXCEPTION 'KNN error: Number of neighbors k must be a positive integer.';
+    END IF;
+    IF (operation != 'c' AND operation != 'r') THEN
+        RAISE EXCEPTION 'KNN error: put r for regression OR c for classification.';
+    END IF;
+    PERFORM MADLIB_SCHEMA.create_schema_pg_temp();
+    EXECUTE
+        $sql$
+    	DROP TABLE IF EXISTS pg_temp.knn_label;
+    	CREATE TABLE pg_temp.knn_label(pid integer, predlabel float);
+	$sql$;
+    
+    --FOR r IN EXECUTE format('SELECT * FROM %I', test_source)
+    FOR r IN EXECUTE format('SELECT %I,%I FROM %I', id_column_name, test_column_name, test_source)
+    LOOP
+        
+	--RAISE NOTICE 'Original: %',r.pid;
+	--RAISE NOTICE 'Original: %',r.p;
+	cur_pid := r.id;
+	vector := r.vector;
+	EXECUTE
+        $sql$
+	DROP TABLE IF EXISTS pg_temp.knn_vector;
+        CREATE TABLE pg_temp.knn_vector(vec DOUBLE PRECISION[]);
+	$sql$;
+	EXECUTE 'INSERT INTO pg_temp.knn_vector values($1)'
+	USING vector;
+	EXECUTE
+        $sql$
+	DROP TABLE IF EXISTS pg_temp.knn_interm;
+        --CREATE TABLE pg_temp.knn_interm(dist DOUBLE PRECISION, lable integer );
+	CREATE TABLE pg_temp.knn_interm AS
+	
+        SELECT madlib.squared_dist_norm2($sql$ || point_column_name || $sql$, vec) as dist, $sql$ || label_column_name || $sql$ FROM $sql$ || textin(regclassout(point_source)) || $sql$, knn_vector order by dist limit $sql$ || k;
+	IF (operation = 'c') THEN
+    	EXECUTE
+        $sql$
+        SELECT mode() within group (order by $sql$ || label_column_name || $sql$) FROM  pg_temp.knn_interm $sql$
+    	INTO l;
+        ELSE
+        EXECUTE
+        $sql$
+        SELECT avg( $sql$ || label_column_name || $sql$ ) FROM  pg_temp.knn_interm $sql$
+        INTO l;
+        END IF;
+	EXECUTE 'INSERT INTO pg_temp.knn_label values($1,$2)'
+	USING cur_pid, l;
+    END LOOP;
+    
+    EXECUTE
+        $sql$
+	DROP TABLE IF EXISTS public.knn_final;
```

Comment:
It might be better not to drop an existing output table. Instead,
we can either take the name of the output table as a parameter,
or we can append the string `"_output"` to the `test_source`
parameter. Of course, we still shouldn't drop the table, if the name
we end up constructing with that append results in a table name
that already exists. The function should throw an error saying the
output table already exists. This behavior will be in line with most
other MADlib functions.